Restore behavior of Box shapes in BulletModel.

avalenzu opened the pull request:
Updating Bullet to 2.87 introduced a fix that caused sub-epsilon
differences in the vertices returned by `btBoxShape::getVertex()`.
SNOPT exhibits different behavior when run with constraints that use the
slightly different set of vertices introduced by the fix.

We were only using `btBoxShape` when adding boxes to the collision model
(the collision shape that was stored in the model was always a
`btConvexHullShape` due to an (unrelated?) issue with collision normals
in `btBoxShape`). It was basically a lazy way to convert box dimensions
to vertex coordinates. This PR explicitly converts the box dimensions to
vertex coordinates and locks down the behavior we expect when creating a
`btCollisionShape` from a `DrakeShapes::Box`.

avalenzu commented:
+@jwnimmer-tri for feature review, please!

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-:-LAItWtFDGur1dwH3-ae:b-qf68jt)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJJQ5bCvBpw6WRzM1K:-LAJJQ5bCvBpw6WRzM1L:b-upiqci) (no related file):*
Needs platform review assigned after fixes.

---

*[multibody/collision/bullet_model.h, line 139 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJIVLj6HzPM7ObE5U1:-LAJIVLj6HzPM7ObE5U2:b-uz6yei) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.h#L139)):*
> ```objc
>       double collision_threshold) override;
> 
>   static std::unique_ptr<btCollisionShape> newBulletBoxShape(
> ```

BTW Do you want a comment here disclaiming support for this as only to be used for testing purposes?  Or would it be okay for non-test code to call this?

---

*[multibody/collision/bullet_model.cc, line 118 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJJ7W01J3f_gTgKg_z:-LAJJ7W2Brx8X8ZO6pY7:b-tw4qoz) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L118)):*
> ```cpp
>    * a btConvexHullShape.
>    */
>   std::unique_ptr<btCollisionShape> bt_shape(new btConvexHullShape());
> ```

FYI If you want, I think this line would work here instead, without any other changes:
```c++
auto bt_shape = std::make_unique<btConvexHullShape>();
```
(The return statement will automatically upcast the unique_ptr.)

Then, you could lose the `dynamic_cast`.

---

*[multibody/collision/test/bullet_model_test.cc, line 12 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJIK-Q-hZjurP8wrL5:-LAJIK-Q-hZjurP8wrL6:bewlezr) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L12)):*
> ```cpp
> namespace collision {
> namespace {
> GTEST_TEST(BulletModelTest, newBulletBoxShapeTest) {
> ```

It would be helpful to have a short sentence or paragraph here explaining what kind of thing this is testing, and why.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-:-LAJJPGQ2VuRkFpIxPrF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Sorry; I saw the subject line and couldn't resist.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[multibody/collision/bullet_model.cc, line 127 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJE53K8w8SNkD2GJzL:-LAJE53K8w8SNkD2GJzM:b-y3nsfd) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L127)):*
> ```cpp
>     for (double y_sign : {1, -1}) {
>       for (double x_sign : {1, -1}) {
>         dynamic_cast<btConvexHullShape*>(bt_shape.get())
> ```

FYI it might lend clarity and compactness if you do a single dynamic cast *outside* the for loop.  You could also create the half-vector size.
e.g.,

```c++
btConvexHullShape& hull = dynamic_cast<btConvexHullShape&>(*bt_shape);
Vector3d half_size = geometry.size / 2;
...
hull.addPoint({x_sign * half_size(0), y_sign * half_size(1), z_sign * half_size(2)});
...
```

---

*[multibody/collision/test/bullet_model_test.cc, line 13 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJG6zQCp8RS0SzJAtc:-LAJG6zQCp8RS0SzJAtd:bgt76ae) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L13)):*
> ```cpp
> namespace {
> GTEST_TEST(BulletModelTest, newBulletBoxShapeTest) {
>   const Vector3<double> half_extents{0.01, 0.04, 0.09};
> ```

FYI I'm curious about these numbers:

0.01 is a number that can't be represented with finite precision in binary.
0.04 is just that value bit-shifted (so, by definition, can't be represented with finite precision). However, it's repeating pattern is the same as 0.01.
0.09 likewise cannot be represented with finite precision but produces a *different* repeating pattern than 0.01 and 0.04.

Is there any particular reason you picked these values?

---

*[multibody/collision/test/bullet_model_test.cc, line 26 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJEuLZ9pu4auVG4fCE:-LAJEuLZ9pu4auVG4fCF:b95s0x1) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L26)):*
> ```cpp
>   };
>   // clang-format on
>   const int expected_num_vertices = expected_vertices.size();
> ```

FYI unsigned value assigned to signed value. Should have a `static_cast`.

---

*[multibody/collision/test/bullet_model_test.cc, line 34 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFjPL1fFKWSUIs84c:-LAJFjPL1fFKWSUIs84d:b-1awex5) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L34)):*
> ```cpp
>   // Lock down implementation as btConvexHullShape as we'll need that to check
>   // the vertices below.
>   ASSERT_TRUE(bt_convex_hull_shape);
> ```

FYI Would this be clearer as:

```c++
ASSERT_NE(bt_convex_hull_shape, nulltpr);
```

---

*[multibody/collision/test/bullet_model_test.cc, line 40 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFFG7-XT-39hSPJ8C:-LAJFFG7-XT-39hSPJ8D:bv4wnce) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L40)):*
> ```cpp
> 
>   // Check that the vertices have the expected values. We are intentionally
>   // comparing doubles here because the results of queries on BulletModel may be
> ```

FYI I think the point that you're making is that you're comparing the doubles to bit-level equivalency.

Given the nature of this test, why not:

```c++
EXPECT_EQ(bullet_vertex.x(), expected_vertices[i].x());
```

Why the subtraction compared to zero?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-:-LAJKFUV2pazwoPEvQGL:b2x7l1h)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/collision/bullet_model.cc, line 127 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJE53K8w8SNkD2GJzL:-LAJKYcd7Rq207yMEIhR:b-d9mqf3) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L127)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI it might lend clarity and compactness if you do a single dynamic cast *outside* the for loop.  You could also create the half-vector size.
e.g.,

```c++
btConvexHullShape& hull = dynamic_cast<btConvexHullShape&>(*bt_shape);
Vector3d half_size = geometry.size / 2;
...
hull.addPoint({x_sign * half_size(0), y_sign * half_size(1), z_sign * half_size(2)});
...
```


</blockquote></details>

I defer to @jwnimmer-tri  on this one.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/collision/bullet_model.cc, line 127 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJE53K8w8SNkD2GJzL:-LAJODKm2L5-91Mdgbjc:b-elowd9) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L127)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I defer to @jwnimmer-tri  on this one.
</blockquote></details>

FYI dynamic_cast is _very_ expensive and you definitely don't want to do it more often than needed. _Possibly_ the compiler would optimize this properly but I wouldn't count on it.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Addressed reviewer comments. @sammy-tri  for platform review.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/collision/bullet_model.h, line 139 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJIVLj6HzPM7ObE5U1:-LAJq4DR1Okh9Tho1DM8:b-37guna) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.h#L139)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Do you want a comment here disclaiming support for this as only to be used for testing purposes?  Or would it be okay for non-test code to call this?
</blockquote></details>

If non-test code wanted to call this, it would be fine. It's static, so its not like it's going to mess with any instances of the class that we care about.

---

*[multibody/collision/bullet_model.cc, line 118 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJJ7W01J3f_gTgKg_z:-LAJqOKHD1ictG72bBvL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI If you want, I think this line would work here instead, without any other changes:
```c++
auto bt_shape = std::make_unique<btConvexHullShape>();
```
(The return statement will automatically upcast the unique_ptr.)

Then, you could lose the `dynamic_cast`.
</blockquote></details>

Done.

---

*[multibody/collision/bullet_model.cc, line 127 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJE53K8w8SNkD2GJzL:-LAJqOwv3I6bZFd7uYIq:b-ewkmqv) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/bullet_model.cc#L127)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI dynamic_cast is _very_ expensive and you definitely don't want to do it more often than needed. _Possibly_ the compiler would optimize this properly but I wouldn't count on it.
</blockquote></details>

Went with @jwnimmer-tri 's suggestion above.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-:-LAJqWAF5DB7J8dQoLo2:blfq388)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/collision/test/bullet_model_test.cc, line 12 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJIK-Q-hZjurP8wrL5:-LAJr7tdEthec2B5CyN3:b-o8lhh0) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It would be helpful to have a short sentence or paragraph here explaining what kind of thing this is testing, and why.
</blockquote></details>

@avalenzu This still needs attention.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Responded to remaining feature reviewer comments. @jwnimmer-tri , @seancurtis-tri , PTAL.

---

Review status: 3 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/collision/test/bullet_model_test.cc, line 13 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJG6zQCp8RS0SzJAtc:-LAJqodu7GJkw5GyyyvJ:b-l7p0jb) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L13)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I'm curious about these numbers:

0.01 is a number that can't be represented with finite precision in binary.
0.04 is just that value bit-shifted (so, by definition, can't be represented with finite precision). However, it's repeating pattern is the same as 0.01.
0.09 likewise cannot be represented with finite precision but produces a *different* repeating pattern than 0.01 and 0.04.

Is there any particular reason you picked these values?
</blockquote></details>

It's the first three squares, in cm. Do you have alternative suggestions? 

---

*[multibody/collision/test/bullet_model_test.cc, line 34 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFjPL1fFKWSUIs84c:-LAKKZEi7X2qLM-NqGif:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L34)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Would this be clearer as:

```c++
ASSERT_NE(bt_convex_hull_shape, nulltpr);
```
</blockquote></details>

Done.

---

*[multibody/collision/test/bullet_model_test.cc, line 40 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFFG7-XT-39hSPJ8C:-LAJrPN149MGoXiUHBKw:b-1ugb60) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L40)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I think the point that you're making is that you're comparing the doubles to bit-level equivalency.

Given the nature of this test, why not:

```c++
EXPECT_EQ(bullet_vertex.x(), expected_vertices[i].x());
```

Why the subtraction compared to zero?
</blockquote></details>

I had the code that you suggested originally, but then the output if the test fails looks like this:

```
multibody/collision/test/bullet_model_test.cc:46: Failure
      Expected: bullet_vertex.x()
      Which is: 0.01
To be equal to: expected_vertices[i].x()
      Which is: 0.01
```
With the subtraction, you get
```
multibody/collision/test/bullet_model_test.cc:46: Failure
      Expected: bullet_vertex.x() - expected_vertices[i].x()
      Which is: 1.73472e-18
To be equal to: 0.
      Which is: 0
```h
```


---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-:-LAKNhz_EBDZ6kXMLs_T:b-ujj7as)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[multibody/collision/test/bullet_model_test.cc, line 13 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJG6zQCp8RS0SzJAtc:-LAKP3ZV1s3C5PrPBRwB:beugm1j) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L13)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

It's the first three squares, in cm. Do you have alternative suggestions? 
</blockquote></details>

Nope. I was just curious. Given that this is all about detecting precision issues down around epsilon, I thought there might be some cunning reasoning.

---

*[multibody/collision/test/bullet_model_test.cc, line 40 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFFG7-XT-39hSPJ8C:-LAKPyAN1jS8SQ-NGPAV:b-advcad) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L40)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I had the code that you suggested originally, but then the output if the test fails looks like this:

```
multibody/collision/test/bullet_model_test.cc:46: Failure
      Expected: bullet_vertex.x()
      Which is: 0.01
To be equal to: expected_vertices[i].x()
      Which is: 0.01
```
With the subtraction, you get
```
multibody/collision/test/bullet_model_test.cc:46: Failure
      Expected: bullet_vertex.x() - expected_vertices[i].x()
      Which is: 1.73472e-18
To be equal to: 0.
      Which is: 0
```h
</blockquote></details>

An excellent point ; worth documenting so people down the road don't ask, "Hur?"

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[multibody/collision/test/bullet_model_test.cc, line 40 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611#-LAJFFG7-XT-39hSPJ8C:-LAKQqYT6XlN9E99pfs0:bju78eg) ([raw file](https://github.com/robotlocomotion/drake/blob/f08734aa33e30cad1c560ec46bab8d29a645da51/multibody/collision/test/bullet_model_test.cc#L40)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

An excellent point ; worth documenting so people down the road don't ask, "Hur?"
</blockquote></details>

FYI Does `EXPECT_NEAR(a, b, 0.0);` give a good output?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8611)*
<!-- Sent from Reviewable.io -->
